### PR TITLE
Get or create token

### PR DESCRIPTION
**Pull request by leplatrem**
Rewrite of #5

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -23,10 +23,13 @@
       }
 
       if (options.credentials) {
-        var hawkHeader = hawk.client.header(options.url, options.method, {
-          credentials: options.credentials
-        });
-        req.setRequestHeader('authorization', hawkHeader.field);
+        try {
```
Comment (Natim) [inline index.js]: You shouldn't have try/catch here it will hide the header already written errors.

File: index.js [hunk 2]
```diff
@@ -49,11 +52,38 @@
     });
   }
 
-  function getToken(daybedUrl) {
-    return request({
-      method: "POST",
-      url: daybedUrl + "/tokens"
-    });
+  function _credentials(credentials) {
```
Comment (Natim) [inline index.js]: Should it be renamed `check_credentials` or `clean_credentials`?

File: index.js [hunk 3]
```diff
@@ -49,11 +52,38 @@
...
+
+  function getToken(host, credentials) {
+    if (host === undefined) {
+      throw new Error("You should provide an host.");
+    }
+    credentials = _credentials(credentials);
+    if (credentials) {
+      return request({
+        method: "GET",
+        url: host + "/token",
```
Comment (Natim) [inline index.js:106]: Nice :+1:

File: index.js [hunk 4]
```diff
@@ -70,23 +100,36 @@
...
   }
 
+  function startSession(host, options) {
+    options = options || {
+      credentials: undefined,
+      getToken: undefined
+    };
+
+    var credentials = options.credentials;
+    var token = options.getToken && options.getToken();
```
Comment (Natim) [inline index.js]: I don't understand the `options.getToken()` here. Shouldn't it raise the `"host" missing` error? Or even return a promise?

File: test/test.session.js [hunk 5]
```diff
@@ -1,9 +1,60 @@
...
+            credentials: {id: 'a', key: 'xyz'},
+        }).then(function (session) {
+            assert.equal(session.credentials.id, '3.14');
+            done();
+        });
+
+        server.respond();
+    });
+
+    it("should derive the token if specified via option", function (done) {
```
Comment (Natim) [inline test/test.session.js]: Could you had a test for when token is not a function

File: index.js (hunk 4 above)
Comment (Natim) [inline index.js]: I think I've got it the `options.getToken` has nothing to do with `getToken`. Should we name it differently then?

File: index.js (hunk 1 above)
Comment (leplatrem) [inline index.js]: Ok got it

File: index.js (hunk 4 above)
Comment (leplatrem) [inline index.js]: `provideToken()` ?

Comment (Natim) [inline index.js]: Yes or `readToken()` maybe?

Comment (leplatrem) [inline index.js]: read is better :)

**Comment by leplatrem** [comment]
Ready to review ! ...and merge :)

File: index.js [hunk 6]
```diff
@@ -26,37 +25,105 @@
         var hawkHeader = hawk.client.header(options.url, options.method, {
           credentials: options.credentials
         });
-        req.setRequestHeader('authorization', hawkHeader.field);
+        req.setRequestHeader('Authorization', hawkHeader.field);
       }
+
       req.onload = function() {
+        // With some browsers, read responseText.
+        // On some server error, maybed no JSON.
```
Comment (almet) [inline index.js]: typo: maybed.

**Pull request merged**